packages/library/src/rpc-request-coalescer.ts

```diff
@@ -0,0 +1,63 @@
+import { IRpcTransport } from '@solana/rpc-transport/dist/types/transports/transport-types';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import fastStableStringify from 'fast-stable-stringify';
+import { shouldCoalesceRequest } from './rpc-request-coalescer-gating';
+
+type CoalescedRequest = {
+    readonly abortController: AbortController;
+    numConsumers: number;
+    readonly responsePromise: Promise<unknown>;
+};
+
+export function getRpcTransportWithRequestCoalescing(transport: IRpcTransport): IRpcTransport {
+    const coalescedRequestsByHash: Record<string, CoalescedRequest> = {};
+    return async function makeCoalescedHttpRequest<TResponse>(
+        config: Parameters<IRpcTransport>[0]
+    ): Promise<TResponse> {
+        const { payload, signal } = config;
+        if (!shouldCoalesceRequest(payload)) {
+            return await transport(config);
+        }
+        const requestHash = fastStableStringify([payload.method, payload.params]);
+        if (coalescedRequestsByHash[requestHash] == null) {
+            const abortController = new AbortController();
+            coalescedRequestsByHash[requestHash] = {
+                abortController,
+                numConsumers: 1,
+                responsePromise: (async () => {
+                    try {
+                        return await transport<TResponse>({
+                            ...config,
+                            signal: abortController.signal,
+                        });
+                    } finally {
+                        delete coalescedRequestsByHash[requestHash];
+                    }
+                })(),
+            };
```

Comment:
Do you not already have an in-flight request for this method/params combo? Make one!